Source/DafnyCore/DafnyOptions.cs

```diff
@@ -1055,35 +1058,85 @@ public enum IncludesModes {
 
     /// <summary>
     /// Dafny releases come with their own copy of Z3, to save users the trouble of having to install extra dependencies.
-    /// For this to work, Dafny looks for Z3 at the location where it is put by our release script (i.e., z3/bin/z3[.exe]).
-    /// If Z3 is not found there, Dafny relies on Boogie to locate Z3 (which also supports setting a path explicitly on the command line).
-    /// Developers (and people getting Dafny from source) need to install an appropriate version of Z3 themselves.
+    /// For this to work, Dafny first tries any prover path explicitly provided by the user, then looks for for the copy
+    /// distributed with Dafny, and finally looks in any directory in the system PATH environment variable.
     /// </summary>
-    private void SetZ3ExecutablePath() {
+    private Version SetZ3ExecutablePath() {
+      string confirmedProverPath = null;
+
+      // Try an explicitly provided prover path, if there is one.
       var pp = "PROVER_PATH=";
       var proverPathOption = ProverOptions.Find(o => o.StartsWith(pp));
       if (proverPathOption != null) {
         var proverPath = proverPathOption.Substring(pp.Length);
         // Boogie will perform the ultimate test to see if "proverPath" is real--it will attempt to run it.
         // However, by at least checking if the file exists, we can produce a better error message in common scenarios.
+        // Unfortunately, there doesn't seem to be a portable way of checking whether it's executable.
         if (!File.Exists(proverPath)) {
-          throw new Bpl.ProverException($"Requested prover not found: '{proverPath}'");
+          return null;
         }
-      } else {
-        var platform = (int)System.Environment.OSVersion.Platform;
 
-        var isUnix = platform == 4 || platform == 6 || platform == 128;
+        confirmedProverPath = proverPath;
+      }
 
-        var z3binName = isUnix ? "z3" : "z3.exe";
+      var platform = System.Environment.OSVersion.Platform;
+      var isUnix = platform == PlatformID.Unix || platform == PlatformID.MacOSX || (int)platform == 128;
```

Comment:
Can you please add a comment about this 128?

Reply:
I honestly don't know what it is. The code previously compared with three constant integers that weren't explained. I saw that two of them were equal to named constants in the `PlatformID` class, so I changed the code to refer to those, but the third one didn't correspond to any documented value. I suspect it actually shows up in practice, but I don't know when. This version seems better than the old code, even if not perfect.

Reply:
I would feel comfortable removing it as, according to the documentation, 128 is not a valid value:
https://learn.microsoft.com/en-us/dotnet/api/system.platformid?view=net-7.0

Reply:
Sounds good to me.